File: Python/Euler03/Euler03_b.py
#The prime factors of 13195 are
# 5, 7, 13 and 29.
#What is the largest prime factor
#of the number 600851475143

# makes a list of prime factors and selects highest
# ~ 190000000 nanoseconds

from math import sqrt
import time

start = time.perf_counter_ns()

def is_prime(no):
    if no < 2:
        return False
    elif no == 2:
        return True
    elif (no & 1) != 1: # bitwise even test
        return False
    for i in range(3, int(sqrt(no)), 2):
        if no % i == 0:
            return False
    return True

current_number = 600851475143
factors = []

for i in range(2, current_number):
    if is_prime(i):
        if current_number % i == 0:
            factors.append(i)
            current_number = int(current_number / i)
    if is_prime(current_number):
        factors.append(current_number)
        break

duration = time.perf_counter_ns() - start

print("Prime factors: ", factors)
print("Largest prime factor: ", end="")
print(max(factors))
print("Took: %s nanoseconds" % (duration))

File: C_Sharp/Euler03/Euler03_b.cs
/* The prime factors of 13195 are 5, 7, 13 and 29.
What is the largest prime factor of the number 600851475143
(Answer = 6857) */

// Count from 1 to target, checking for factors which are prime, add each prime factor to a list,
// find product of list, if product = target, last prime added is highest 
// ~ 13000 ticks

using System;
using System.Collections.Generic;
using System.Diagnostics;

namespace Euler03_bname
{
    class Euler03_b
    {
        static Boolean isPrime(int target)
        {
            if (target <= 1) return false; // 1 isnt prime
            if (target == 2) return true; // 2 is prime
            if ((target & 1) != 1) return false; // bitwise even test
            var boundary = (int)Math.Floor(Math.Sqrt(target));
            for (int i = 3; i <= boundary; i += 2) // increment thru odds
            {
                if (target % i == 0) return false;
            }
            return true;
        }

        static void Main(string[] args)
        {
            var watch = new System.Diagnostics.Stopwatch();
            long num = 600851475143;
            List<long> pflist = new List<long>(); // list to store prime factors
            long prod = 1; // store product of prime factors
            watch.Start();
            for (int i = 1; i < num; i++)
            {
                if (num % i == 0 && isPrime(i) == true) // check if prime factor
                {   
                    pflist.Add(i); // add to list
                    prod = 1; // reset product (must be at least 1)
                    foreach (var m in pflist) { prod *= m; } // multiply list contents   
                }
                if (prod == num) // if product = target, prime factor is highest
                {
                    Console.WriteLine("Answer: {0}", i);               
                    break; // stop the for loop early
                }
            }
            watch.Stop();
            Console.WriteLine("Took: {0} ticks", watch.ElapsedTicks);
        }
    }
}
